src/OpenTelemetry/Metrics/MetricReaderExt.cs

```diff
@@ -252,6 +237,52 @@ internal void SetMaxMetricPointsPerMetricStream(int maxMetricPointsPerMetricStre
         }
     }
 
+    private static void DeactivateMetric(Metric metric)
+    {
+        metric.InstrumentDisposed = true;
+
+        var metricStreamIdentity = metric.InstrumentIdentity;
+
+        OpenTelemetrySdkEventSource.Log.MetricInstrumentDeactivated(
+            metricStreamIdentity.InstrumentName,
+            metricStreamIdentity.MeterName);
+    }
+
+    private void CreateOrUpdateMetricStreamRegistration(in MetricStreamIdentity metricStreamIdentity)
+    {
+        var registration = this.metricStreamNames.GetOrAdd(
+            metricStreamIdentity.MetricStreamName,
+            CreateRegistration);
+
+        var currentRegistrationCount = Interlocked.CompareExchange(ref registration.RegistrationCount, 1, -1);
+
+        if (currentRegistrationCount == -1)
+        {
+            // Most common case where instrument being added is the first registration.
+            return;
+        }
+
+        if (currentRegistrationCount > 0)
+        {
+            OpenTelemetrySdkEventSource.Log.DuplicateMetricInstrument(
+                metricStreamIdentity.InstrumentName,
+                metricStreamIdentity.MeterName,
+                "Metric instrument has the same name as an existing one but differs by description, unit, or instrument type. Measurements from this instrument will still be exported but may result in conflicts.",
+                "Either change the name of the instrument or use MeterProviderBuilder.AddView to resolve the conflict.");
+        }
+        else
+        {
+            OpenTelemetrySdkEventSource.Log.MetricInstrumentReactivated(
```

Comment:
I think we should rely on `MetricStreamIdentity` instead of `MetricStreamIdentity.MetricStreamName` to decide reactivation/deactivation.

Reply:
It's related to https://github.com/open-telemetry/opentelemetry-dotnet/pull/4958/files#r1396717590